### DISCUSSION TITLE
Multi-root with CCLS

### DISCUSSION THREAD

**Discussion by spaceotter** [emacs-lsp/lsp-mode#4249] [Q&A]
I have an (annoying) project setup with two folders, say A and B, A has the build folder with the compilation_commands and is set up and working with CCLS. There is a separate folder B which contains some files that are part of the project. The command `lsp-workspace-folders-add` does not do the desired thing when I add folder B. This causes it to start a separate server with the incorrect root. If I invoke `lsp` manually, I can select the correct root A and it connects to the right server. How can I get files in B to automatically connect to the server for A?

**Comment by yyoncho** [comment]
Something like this should work for you:

```
(setq ccls-library-folders-fn (lambda (&rest _)
                                (list "/folders/that/should/be/considered/part/of/the/main/project")))
```

**Reply by spaceotter** [reply]
That doesn't seem to do anything.

**Reply by spaceotter** [reply]
The documentation looks promising but I'm really struggling to find what the variable actually changed.

**Reply by yyoncho** [reply]
Have you removed the roots that you have already added via M-x lsp-workspace-folders-remove?

**Reply by spaceotter** [reply]
Yes, I did.

**Reply by yyoncho** [reply]
I expect the following: You have the root of the project /foo which contains the core files. You have /bar which contains library files that are locally part of the project but not physically (by path). The first one should be the project root (visible via M-x lsp-describe-session) the later one should be in the list as described here: https://github.com/emacs-lsp/lsp-mode/discussions/4249#discussioncomment-7726210 .

Then when you open a file under /foo it should start a server connection. When you then open a file under bar it should use the same server connection. Let me know if this is what you want and what are your observations when you perform the steps.

**Reply by spaceotter** [reply]
Yes, that is the behavior I want.
The behavior I see is that opening a file under bar doesn't start or use any server connection.

**Reply by yyoncho** [reply]
> The behavior I see is that opening a file under bar doesn't start or use any server connection.

And what do you see when you do M-x lsp in a file under bar?

**Reply by spaceotter** [reply]
Sorry, was busy with something else and I just got back around to this.
When I do M-x lsp in a file under bar/, it tells me the file "is not part of any project" and asks to import a project root.

**Reply by spaceotter** [reply]
The menu lets me manually select /foo as the root. This doesn't work, it connects to the same ccls process as the files in /foo but ccls doesn't seem to be picking up the build arguments from /foo's compile_commands.json.

**Reply by spaceotter** [reply]
So, I modified ccls.el to set the client :library-folders-fn to the lambda directly.
Files under bar/ now connect to the foo/ server, but again don't pick up the build arguments.
